File: p2_odometry/l3_intro/src/bumperbot_py_examples/bumperbot_py_examples/simple_subscriber.py
import rclpy  # Nhập thư viện ROS 2 chính cho Python để quản lý vòng đời và giao tiếp
from rclpy.node import Node  # Nhập lớp Node để kế thừa và tạo ra các thành phần ROS 2
from std_msgs.msg import String  # Nhập kiểu dữ liệu chuỗi văn bản chuẩn (String) để nhận gói tin dữ liệu


class SimpleSubscriber(Node):  # Định nghĩa lớp SimpleSubscriber kế thừa từ lớp Node của ROS 2

    def __init__(self):  # Hàm khởi tạo (constructor) chạy ngay khi đối tượng được tạo từ lớp này
        super().__init__("simple_subscriber")  # Gọi hàm khởi tạo lớp cha và đặt tên Node này là "simple_subscriber"
        self.sub_ = self.create_subscription(String, "chatter", self.msgCallback, 10)  # Tạo Subscriber lắng nghe kiểu String trên topic "chatter", gọi hàm msgCallback khi có tin và hàng đợi = 10
        self.sub_  # Dòng này dùng để giữ tham chiếu đến đối tượng subscription (tránh bị bộ gom rác Python hiểu nhầm và xóa đi)

    def msgCallback(self, msg):  # Định nghĩa hàm callback, tự động chạy và nhận tham số 'msg' mỗi khi có tin nhắn mới truyền tới topic
        self.get_logger().info("I heard: %s" % msg.data)  # Trích xuất chuỗi từ trường dữ liệu '.data' của tin nhắn và in log ra màn hình console


def main():  # Định nghĩa hàm main - điểm khởi đầu chính của chương trình
    rclpy.init()  # Khởi tạo các cấu hình nền tảng và kết nối của thư viện rclpy

    simple_publisher = SimpleSubscriber()  # Tạo một thực thể (instance) cụ thể cho Node SimpleSubscriber (tên biến đang đặt tạm là simple_publisher)
    rclpy.spin(simple_publisher)  # Giữ Node chạy liên tục trong vòng lặp vô hạn để sẵn sàng đón nhận và xử lý tin nhắn từ topic

    simple_publisher.destroy_node()  # Giải phóng tài nguyên và hủy Node sau khi thoát khỏi vòng lặp spin (khi nhấn Ctrl+C)
    rclpy.shutdown()  # Tắt hoàn toàn thư viện rclpy và ngắt kết nối hệ thống với mạng ROS 2


if __name__ == '__main__':  # Kiểm tra xem file script này có đang được thực thi trực tiếp từ dòng lệnh hay không
    main()  # Gọi hàm main để bắt đầu chạy toàn bộ chương trình Node Subscriber

File: p2_odometry/src/bumperbot_cpp_examples/src/simple_subscriber.cpp
#include <rclcpp/rclcpp.hpp>  // Nhập thư viện client chính của ROS 2 dành cho C++ (quản lý Node, Subscriber...)
#include <std_msgs/msg/string.hpp>  // Nhập kiểu dữ liệu chuỗi văn bản tiêu chuẩn (String) trong hệ thống ROS 2


using std::placeholders::_1;  // Khai báo placeholder _1 để đại diện cho tham số đầu tiên (gói tin nhắn) truyền vào hàm callback khi dùng std::bind

class SimpleSubscriber : public rclcpp::Node  // Định nghĩa lớp SimpleSubscriber kế thừa từ lớp rclcpp::Node của ROS 2
{
public:
  SimpleSubscriber() : Node("simple_subscriber")  // Hàm khởi tạo (constructor): Gọi hàm khởi tạo lớp cha và đặt tên Node là "simple_subscriber"
  {
    sub_ = create_subscription<std_msgs::msg::String>(  // Khởi tạo một Subscriber lắng nghe kiểu dữ liệu gói tin dạng String
        "chatter", 10, std::bind(&SimpleSubscriber::msgCallback, this, _1));  // Đăng ký nghe trên topic "chatter", kích thước hàng đợi = 10, và liên kết sự kiện tới hàm msgCallback với 1 tham số dữ liệu đầu vào (_1)
  }

private:
  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub_;  // Khai báo con trỏ thông minh (SharedPtr) để quản lý đối tượng Subscriber lắng nghe kiểu tin nhắn String

  void msgCallback(const std_msgs::msg::String &msg) const  // Định nghĩa hàm callback (hàm hằng - const) tự động kích hoạt để xử lý gói tin tham chiếu 'msg' mỗi khi có tin nhắn mới đổ về
  {
    RCLCPP_INFO_STREAM(this->get_logger(), "I heard: " << msg.data.c_str());  // In dòng log dạng stream cấp độ INFO ra console, trích xuất chuỗi văn bản bằng phương thức .c_str() của C++
  }
};


int main(int argc, char * argv[])  // Hàm main - điểm khởi đầu và thực thi chính của chương trình C++
{
  rclcpp::init(argc, argv);  // Khởi tạo các cấu hình hệ thống, thiết lập lớp nền tảng và phân tích tham số dòng lệnh cho ROS 2
  rclcpp::spin(std::make_shared<SimpleSubscriber>());  // Tạo nhanh một con trỏ thông minh cho Node SimpleSubscriber và đưa vào vòng lặp vô hạn (spin) để lắng nghe và gọi callback xử lý dữ liệu
  rclcpp::shutdown();  // Tắt hoàn toàn môi trường ROS 2 và giải phóng tài nguyên sau khi kết thúc vòng lặp spin (khi bấm Ctrl+C)
  return 0;  // Kết thúc chương trình chính và trả về mã trạng thái thành công 0 cho hệ điều hành
}